qsc/plot.py

```diff
@@ -0,0 +1,529 @@
+    z_middle = np.mean(z_limits)
+
+    # The plot bounding box is a sphere in the sense of the infinity
+    # norm, hence call half the max range the plot radius.
+    plot_radius = 0.5 * max([x_range, y_range, z_range])
+
+    ax.set_xlim3d([x_middle - plot_radius, x_middle + plot_radius])
+    ax.set_ylim3d([y_middle - plot_radius, y_middle + plot_radius])
+    ax.set_zlim3d([z_middle - plot_radius, z_middle + plot_radius])
+
+def create_subplot(ax, x_2D_plot, y_2D_plot, z_2D_plot, colormap, elev=90, azim=45, dist=7, **kwargs):
+    '''
+    Construct the surface given a surface in cartesian coordinates
+    x_2D_plot, y_2D_plot, z_2D_plot already with phi=[0,2*pi].
+    A matplotlib figure with elements fig, ax
+    must have been previously created.
+
+    Args:
+        ax: matplotlib figure instance
+        x_2d_plot: 2D array for the x coordinates of the surface
+        y_2d_plot: 2D array for the x coordinates of the surface
+        z_2d_plot: 2D array for the x coordinates of the surface
+        elev: elevation angle for the camera view
+        azim: azim angle for the camera view
+        distance: distance parameter for the camera view
+    '''
+    ax.plot_surface(x_2D_plot, y_2D_plot, z_2D_plot, facecolors=colormap,
+                    rstride=1, cstride=1, antialiased=False,
+                    linewidth=0, alpha=1., shade=False, **kwargs)
+    set_axes_equal(ax)
+    ax.set_axis_off()
+    ax.dist = dist
+    ax.elev = elev
+    ax.azim = azim
+
+def create_field_lines(qsc, alphas, X_2D, Y_2D, Z_2D, phimax=2*np.pi, nphi=500):
+    '''
+    Function to compute the (X, Y, Z) coordinates of field lines at
+    several alphas, where alpha = theta-iota*varphi with (theta,varphi)
+    the Boozer toroidal angles. This function relies on a 2D interpolator
+    from the scipy library to smooth out the lines
+
+    Args:
+      qsc: instance of self
+      alphas: array of field line labels alpha
+      X_2D: 2D array for the x components of the surface
+      Y_2D: 2D array for the y components of the surface
+      Z_2D: 2D array for the z components of the surface
+      phimax: maximum value for the field line following angle phi
+      nphi: grid resolution for the output fieldline
+    '''
+    phi_array = np.linspace(0,phimax,nphi,endpoint=False)
+    fieldline_X = np.zeros((len(alphas),nphi))
+    fieldline_Y = np.zeros((len(alphas),nphi))
+    fieldline_Z = np.zeros((len(alphas),nphi))
+    [ntheta_RZ,nphi_RZ] = X_2D.shape
+    phi1D   = np.linspace(0,2*np.pi,nphi_RZ)
+    theta1D = np.linspace(0,2*np.pi,ntheta_RZ)
+    X_2D_spline = interp2d(phi1D, theta1D, X_2D, kind='cubic')
+    Y_2D_spline = interp2d(phi1D, theta1D, Y_2D, kind='cubic')
+    Z_2D_spline = interp2d(phi1D, theta1D, Z_2D, kind='cubic')
+    for i in range(len(alphas)):
+        for j in range(len(phi_array)):
+            phi_mod = np.mod(phi_array[j],2*np.pi)
+            varphi0=qsc.nu_spline(phi_array[j])+2*phi_array[j]-phi_mod
```

Comment:
I'm not clear why this line is not `varphi0 = qsc.nu_spline(phi_array[j]) + phi_array[j]`.

Reply:
The tricky thing here is nu_spline. If phi_array has a maximum that extends beyond 2pi, then nu_spline only gives the nu part corresponding to an argument mod(phi,2*np.pi). The correct value for nu is obtained by setting nu = nu_spline + phi_array - phi_med as nu is periodic.